signer/signer_trust.go

```diff
@@ -85,3 +95,56 @@ func (trust *NotarySigner) GetKey(keyid string) data.PublicKey {
 	}
 	return data.NewPublicKey(data.KeyAlgorithm(publicKey.KeyInfo.Algorithm.Algorithm), publicKey.PublicKey)
 }
+
+// CheckHealth returns true if trust is healthy - that is if it can connect
+// to the trust server, and both the key management and signer services are
+// healthy
+func (trust *NotarySigner) CheckHealth(timeout int) error {
+	if e := trust.checkServiceHealth(
+		"Key manager", trust.kmClient.CheckHealth, timeout); e != nil {
+		return e
+	}
+	return trust.checkServiceHealth(
+		"Signer", trust.sClient.CheckHealth, timeout)
+}
+
+type rpcHealthCheck func(
+	context.Context, *pb.Void, ...grpc.CallOption) (*pb.HealthStatus, error)
+
+// Generalized function that can check the health of both the signer client
+// and the key management client.
+func (trust *NotarySigner) checkServiceHealth(
+	serviceName string, check rpcHealthCheck, timeout int) error {
+
+	// Do not bother starting goroutine if the connection is broken.
+	if trust.clientConn.State() != grpc.Idle &&
+		trust.clientConn.State() != grpc.Ready {
+		return errors.New("Not currently connected to trust server.")
+	}
+
+	// We still want to time out getting health, because the connection could
+	// have disconnected sometime between when we checked the connection and
+	// when we try to make an RPC call.
+	channel := make(chan error)
+	done := make(chan struct{})
+	go func() {
+		status, err := check(context.Background(), &pb.Void{})
+		if len(status.Status) > 0 {
+			err = fmt.Errorf("%s not healthy", serviceName)
+		}
+		select {
+		case <-done:
+		case channel <- err:
+		}
```

Comment:
Another way to do this would be to make `channel` a buffered channel (`make(chan error, 1)`). Then instead of this select statement, you could just do `channel <- err`. If the timeout has already elapsed and the outer function has returned, it's fine, because the channel is buffered so the write to it won't block. This would make the `done` channel unnecessary.